internal/oidc/auth/auth_handler.go

```diff
@@ -43,107 +50,201 @@ func NewHandler(
 			return httperr.Newf(http.StatusMethodNotAllowed, "%s (try GET or POST)", r.Method)
 		}
 
-		csrfFromCookie := readCSRFCookie(r, cookieCodec)
-
-		authorizeRequester, err := oauthHelper.NewAuthorizeRequest(r.Context(), r)
-		if err != nil {
-			plog.Info("authorize request error", oidc.FositeErrorForLog(err)...)
-			oauthHelper.WriteAuthorizeError(w, authorizeRequester, err)
-			return nil
-		}
-
-		upstreamIDP, err := chooseUpstreamIDP(idpListGetter)
+		oidcUpstream, ldapUpstream, err := chooseUpstreamIDP(idpLister)
```

Comment:
It would be nice to reframe this as "how you authenticate" instead of "what IDP type are you authenticating to."

For example, OIDC can involve username/password or a redirect.  Same with OAuth IDPs.

Today, LDAP only involves username/password, in the future it may involve redirects or Kerberos.

Reply:
Not sure if I fully understood the suggestion. I'll try to respond but let me know if I missed something.

In addition to deciding which method of authentication to use (e.g. password vs kerberos vs redirect), the server also needs to select which upstream IDP to use to authenticate your request, especially when we start supporting allowing multiple upstream IDPs to be configured for a single `FederationDomain`. Since this is all private implementation detail inside the handler, we can refactor it in the future when we want to support more methods of authentication (e.g. password vs kerberos vs redirect).